crates/chia-consensus/src/gen/condition_tools.rs

```diff
@@ -0,0 +1,136 @@
+use crate::gen::opcodes::{
+    ConditionOpcode, AGG_SIG_AMOUNT, AGG_SIG_ME, AGG_SIG_PARENT, AGG_SIG_PARENT_AMOUNT,
+    AGG_SIG_PARENT_PUZZLE, AGG_SIG_PUZZLE, AGG_SIG_PUZZLE_AMOUNT,
+};
+use crate::gen::owned_conditions::{OwnedSpend, OwnedSpendBundleConditions};
+use crate::gen::validation_error::ErrorCode;
+use chia_bls::PublicKey;
+use chia_protocol::Bytes;
+use chia_protocol::Coin;
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+
+pub fn pkm_pairs(
+    conditions: OwnedSpendBundleConditions,
+    additional_data: &[u8],
+) -> Result<(Vec<PublicKey>, Vec<Vec<u8>>), ErrorCode> {
+    let mut pks = Vec::<PublicKey>::new();
+    let mut msgs = Vec::<Vec<u8>>::new();
+    let disallowed = agg_sig_additional_data(additional_data);
+    for (pk, msg) in conditions.agg_sig_unsafe {
+        pks.push(pk);
+        msgs.push(msg.as_slice().to_vec());
+        for (_, disallowed_val) in disallowed.clone().into_iter() {
+            if msg.ends_with(disallowed_val.as_slice()) {
+                return Err(ErrorCode::InvalidCondition);
+            }
+        }
+    }
+    for spend in conditions.spends {
+        let condition_items_pairs = [
+            (AGG_SIG_PARENT, spend.clone().agg_sig_parent),
+            (AGG_SIG_PUZZLE, spend.clone().agg_sig_puzzle),
+            (AGG_SIG_AMOUNT, spend.clone().agg_sig_amount),
+            (AGG_SIG_PUZZLE_AMOUNT, spend.clone().agg_sig_puzzle_amount),
+            (AGG_SIG_PARENT_AMOUNT, spend.clone().agg_sig_parent_amount),
+            (AGG_SIG_PARENT_PUZZLE, spend.clone().agg_sig_parent_puzzle),
+            (AGG_SIG_ME, spend.clone().agg_sig_me),
```

Comment:
I don't think you should clone the `Spend` here. just store references in the array